src/RevCPU.cc

```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
-  Threads.at(ThreadIDToSetup)->GetRegFile()->SetX(RevReg::a1, Mem->GetStackTop() + 60);
+  // setup argc
+  RegFile->SetX(RevReg::a0, Argv.size());
+  RegFile->SetX(RevReg::a1, Mem->GetStackTop() + 60);
+  return;
 }
 
 // Checks core 'i' to see if it has any available harts to assign work to
-// if it does and there is work to assign (ie. ThreadQueue is not empty)
+// if it does and there is work to assign (ie. ReadyThreads is not empty)
 // assign it and enable the processor if not already enabled.
 void RevCPU::UpdateThreadAssignments(uint32_t ProcID){
-  // print the thread queue
-  // Get utilization info
-  double Util = Procs[ProcID]->GetHartUtilization();
-  //if( Util > 0.0 ){
-    output.verbose(CALL_INFO, 11, 0, "Core %" PRIu32 " utilization: %.2f%%\n", ProcID, Util);
-  //}
-  // Check if we have room to schedule another thread
-  if( Util < 100  ){
-    output.verbose(CALL_INFO, 10, 0, "Core %" PRIu32 " utilization: %.2f%%\n", ProcID, Util);
-    // We can schedule another thread
-    // Check if we have any threads to schedule
-    if( ThreadQueue.size() ){
-      // Add to this proc's thread list
-      Threads.at(ThreadQueue.front())->SetState(ThreadState::RUNNING);
-      AssignThread(ThreadQueue.front(), ProcID);
-      // AssignedThreads.at(ProcID).emplace_back(Threads.at(ThreadQueue.front()));
-      // output.verbose(CALL_INFO, 6, 1, "Assigning Thread %u to Core %u\n", ThreadQueue.front(), ProcID);
-      // Remove from thread queue
-      ThreadQueue.erase(ThreadQueue.begin());
-      // If this Proc was previously disabled, enable it
-      // if( !Enabled[i] ){ Enabled[i] = true; }
-      Enabled[ProcID] = true;
-    }
-  } // Utilization is 100%, so change nothing
+  // Check if we have anything to assign
+  if( ReadyThreads.empty() ){
+    return;
+  }
+
+  // There is work to assign, check if this proc has room
+  if( Procs[ProcID]->HasIdleHart() ){
+    // There is room, assign a thread
+    // Get the next thread to assign
+    // TODO: Verify this is the correct way to transfer a unique_ptr
+    // Assign the thread
+    Procs[ProcID]->AssignThread(std::move(ReadyThreads.front()));
+
+    // Remove thread from ready threads vector
+    ReadyThreads.erase(ReadyThreads.begin());
+
+    // If this Proc was previously disabled, enable it
+    if( !Enabled[ProcID] ){ Enabled[ProcID] = true; }
+  }
   return;
 }
 
 // Checks for state changes in the threads of a given processor index 'i'
 // and handle appropriately
-void RevCPU::CheckForThreadStateChanges(uint32_t ProcID){
+void RevCPU::HandleThreadStateChangesForProc(uint32_t ProcID){
   // Handle any thread state changes for this core
   // NOTE: At this point we handle EVERY thread that changed state every cycle
-  while( !Procs[ProcID]->GetThreadsThatChangedState().empty() ){
-    auto& Thread = Procs[ProcID]->GetThreadsThatChangedState().front();
+  std::queue<std::unique_ptr<RevThread>> ThreadsThatChangedState = Procs[ProcID]->TransferThreadsThatChangedState();
```

Comment:
`std::queue<std::unique_ptr<RevThread>> ThreadsThatChangedState = ...` might be simpler as just `auto ThreadsThatChangedState = ...`